scipy/spatial/distance.py

```diff
@@ -2350,7 +2547,42 @@ def num_obs_y(Y):
     return d
 
 
-def cdist(XA, XB, metric='euclidean', **kwargs):
+def _prepare_out_argument(out, dtype, expected_shape):
+    if out is None:
+        return np.empty(expected_shape, dtype=dtype)
+
+    if out.shape != expected_shape:
+        raise ValueError("Output array has incorrect shape.")
+    if not out.flags.c_contiguous:
+        raise ValueError("Output array must be C-contiguous.")
+    if out.dtype != np.double:
+        raise ValueError("Output array must be double type.")
+    return out
+
+
+def _pdist_callable(X, *, out, metric, **kwargs):
+    n = X.shape[0]
+    out_size = (n * (n - 1)) // 2
+    dm = _prepare_out_argument(out, np.double, (out_size,))
+    k = 0
+    for i in range(X.shape[0] - 1):
+        for j in range(i + 1, X.shape[0]):
+            dm[k] = metric(X[i], X[j], **kwargs)
+            k += 1
+    return dm
+
+
+def _cdist_callable(XA, XB, *, out, metric, **kwargs):
+    mA = XA.shape[0]
+    mB = XB.shape[0]
+    dm = _prepare_out_argument(out, np.double, (mA, mB))
+    for i in range(mA):
+        for j in range(mB):
+            dm[i, j] = metric(XA[i], XB[j], **kwargs)
```

Comment:
I hadn't realized there was pure Python management of the iteration like this--are you planning to move that out of Python as well? Anyway, that hasn't changed before/after here.

Reply:
Is this as bad as it looks? Every row in each input of points has to be traversed in raw Python then calling into the C function for all those pairs? Or only happening for some metrics/situations?

Reply:
This is meant to allow user-provided metric-functions. In which case I don't think there's any simple way to speed it up. Really it's the API that would need to change for that to work.